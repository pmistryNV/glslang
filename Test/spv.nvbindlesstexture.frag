#version 450
#extension GL_NV_bindless_texture : enable
#extension GL_ARB_gpu_shader_int64 : enable

// Test case 1
// If both bindless_sampler and bound_sampler, or bindless_image and bound_image, are declared at 
// global scope in any compilation unit, a link-time error will be generated
layout (location=0) uniform sampler2D theSampler; // decoration should be BindlessSamplerNV
layout (location=1) uniform layout(rgba32f) image2D theImage1;// decoration should be BindlessImageNV

in vec2 tCoord;

layout(binding = 2) uniform UBO
{
    int some_int;
    float fixed_array[3];
    sampler2D uboSampler[3];
    layout(rgba32f) image2D uboImage;
    uint64_t textureHandle;
    uint64_t textureHandleI;
};

layout(std430, binding = 0) buffer SSBO
{
    int var1;
    float ssboArray[3];
    sampler2D ssboSampler[3];
    layout(rgba32f) image2D inUniformImage;
    uint64_t utextureHandle;
};

// 3. Enables sampler/image type as part of a structure
struct Light
{
   sampler2D theStructSampler;
   layout(rgba32f) image2D inImage;
} test;

void testSamplerConstructors()
{
    sampler1D samp1D = sampler1D(utextureHandle);
    isampler2D isamp2D = isampler2D(utextureHandle);
    usampler3D usamp3D = usampler3D(utextureHandle);
    samplerCube sampCube = samplerCube(utextureHandle); 
    isampler1DArray isamp1DA = isampler1DArray(utextureHandle);
    usampler2DArray usamp2DA = usampler2DArray(utextureHandle);
    isamplerCubeArray isampCubeA = isamplerCubeArray(utextureHandle);

    sampler1DShadow samp1Ds = sampler1DShadow(utextureHandle);
    sampler2DShadow samp2Ds = sampler2DShadow(utextureHandle);
    samplerCubeShadow sampCubes = samplerCubeShadow(utextureHandle);
    sampler1DArrayShadow samp1DAs = sampler1DArrayShadow(utextureHandle);
    sampler2DArrayShadow samp2DAs = sampler2DArrayShadow(utextureHandle);
    samplerCubeArrayShadow sampCubeAs = samplerCubeArrayShadow(utextureHandle);

    // identity constructors
    sampler1D samp1D_iden = sampler1D(samp1D);
    isampler2D isamp2D_iden = isampler2D(isamp2D);
    usampler3D usamp3D_iden = usampler3D(usamp3D);
    samplerCube sampCube_iden = samplerCube(sampCube); 
    isampler1DArray isamp1DA_iden = isampler1DArray(isamp1DA);
    usampler2DArray usamp2DA_iden = usampler2DArray(usamp2DA);
    isamplerCubeArray isampCubeA_iden = isamplerCubeArray(isampCubeA);

    sampler1DShadow samp1Ds_iden = sampler1DShadow(samp1Ds);
    sampler2DShadow samp2Ds_iden = sampler2DShadow(samp2Ds);
    samplerCubeShadow sampCubes_iden = samplerCubeShadow(sampCubes);
    sampler1DArrayShadow samp1DAs_iden = sampler1DArrayShadow(samp1DAs);
    sampler2DArrayShadow samp2DAs_iden = sampler2DArrayShadow(samp2DAs);
    samplerCubeArrayShadow sampCubeAs_iden = samplerCubeArrayShadow(sampCubeAs);

}

void testImageConstructors() {
    // Test additional image layouts
    // uint64_t <-> image Type
    layout(rgba32f) image1D  i1D = layout(rgba32f) image1D(textureHandle);
    layout(rgba32f) image2D  i2D = layout(rgba32f) image2D(textureHandle);
    layout(rgba32f) image3D  i3D = layout(rgba32f) image3D(textureHandle);
    layout(rgba32f) imageCube  iCube = layout(rgba32f) imageCube(textureHandle);

    layout(rgba32f) imageCubeArray  iCubeArray = layout(rgba32f) imageCubeArray(utextureHandle);
    layout(rgba32f) image1DArray    i1DArray = layout(rgba32f) image1DArray(utextureHandle);
    layout(rg16)    image2DArray    i2DArray = layout(rg16)    image2DArray(utextureHandle);
    layout(rgba32f) imageBuffer     iBuffer  = layout(rgba32f) imageBuffer(textureHandle);
    layout(rgba32f) image2DMS       i2DMS = layout(rgba32f) image2DMS(textureHandle);
    layout(rgba32f) image2DMSArray  i2DMSArray = layout(rgba32f) image2DMSArray(utextureHandle);

    layout(r32i) iimage1D  ii1D = layout(r32i) iimage1D(textureHandle);
    layout(r32ui) uimage2D ui2D = layout(r32ui) uimage2D(utextureHandle);
    layout(r32i)  iimage2DMS ii2DMS = layout(r32i)  iimage2DMS(textureHandle);
    layout(r32ui) uimage2DMSArray ui2DMSArray = layout(r32ui) uimage2DMSArray(utextureHandle);

    // Test identity constructors
    layout(rgba32f) image1D  i1D_iden = layout(rgba32f) image1D(i1D);
    layout(rgba32f) image2D  i2D_iden = layout(rgba32f) image2D(i2D);
    layout(rgba32f) image3D  i3D_iden = layout(rgba32f) image3D(i3D);
    layout(rgba32f) imageCube  iCube_iden = layout(rgba32f) imageCube(iCube);

    layout(rgba32f) imageCubeArray  iCubeArray_iden = layout(rgba32f) imageCubeArray(iCubeArray);
    layout(rgba32f) image1DArray    i1DArray_iden = layout(rgba32f) image1DArray(i1DArray);
    layout(rg16)    image2DArray    i2DArray_iden = layout(rg16)    image2DArray(i2DArray);
    layout(rgba32f) imageBuffer     iBuffer_iden  = layout(rgba32f) imageBuffer(iBuffer);
    layout(rgba32f) image2DMS       i2DMS_iden = layout(rgba32f) image2DMS(i2DMS);
    layout(rgba32f) image2DMSArray  i2DMSArray_iden = layout(rgba32f) image2DMSArray(i2DMSArray);

    layout(r32i) iimage1D  ii1D_iden = layout(r32i) iimage1D(ii1D);
    layout(r32ui) uimage2D ui2D_iden = layout(r32ui) uimage2D(ui2D);
    layout(r32i)  iimage2DMS ii2DMS_iden = layout(r32i)  iimage2DMS(ii2DMS);
    layout(r32ui) uimage2DMSArray ui2DMSArray_iden = layout(r32ui) uimage2DMSArray(ui2DMSArray);

}

void main()
{

    sampler2D theLocalSampler = sampler2D(textureHandle);
    textureQueryLevels(theSampler);
    textureQueryLevels(theLocalSampler);
    layout(rgba32f) image2D localimg = layout(rgba32f) image2D(textureHandleI);
    test.inImage = layout(rgba32f) image2D(textureHandleI);

    vec4 res = texture(sampler2D(textureHandle), vec2(0));
    vec4 res1 = imageLoad(localimg, ivec2(0));
    vec4 res2 = imageLoad(layout(rgba32f) image2D(textureHandleI), ivec2(0));


    sampler2D utheSampler = sampler2D(textureHandle);      // uint64 -> sampler
    layout(rgba32f) image2D ulocalImage = layout(rgba32f) image2D(textureHandle); // uint64 -> image
    uint64_t lu64utextureHandle = uint64_t(uboSampler[0]);      // sampler -> uint64
    uint64_t lu64imgutextureHandle = uint64_t(uboImage);	    // image -> uint64
    
    testSamplerConstructors();
    testImageConstructors();
}
