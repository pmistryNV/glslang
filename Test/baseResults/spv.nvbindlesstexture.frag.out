spv.nvbindlesstexture.frag
// Module Version 10000
// Generated by (magic number): 8000a
// Id's are bound by 333

                              Capability Shader
                              Capability Int64
                              Capability StorageImageMultisample
                              Capability ImageCubeArray
                              Capability Sampled1D
                              Capability Image1D
                              Capability SampledCubeArray
                              Capability ImageBuffer
                              Capability ImageMSArray
                              Capability StorageImageExtendedFormats
                              Capability ImageQuery
                              Capability CapabilityBindlessTextureNV
                              Extension  "SPV_NV_bindless_texture"
               1:             ExtInstImport  "GLSL.std.450"
                              MemoryModel Logical GLSL450
                              SamplerImageAddressingModeNV 64
                              EntryPoint Fragment 4  "main" 332
                              ExecutionMode 4 OriginUpperLeft
                              Source GLSL 450
                              SourceExtension  "GL_ARB_gpu_shader_int64"
                              SourceExtension  "GL_NV_bindless_texture"
                              Name 4  "main"
                              Name 6  "testSamplerConstructors("
                              Name 8  "testImageConstructors("
                              Name 14  "samp1D"
                              Name 24  "SSBO"
                              MemberName 24(SSBO) 0  "var1"
                              MemberName 24(SSBO) 1  "ssboArray"
                              MemberName 24(SSBO) 2  "ssboSampler"
                              MemberName 24(SSBO) 3  "inUniformImage"
                              MemberName 24(SSBO) 4  "utextureHandle"
                              Name 26  ""
                              Name 35  "isamp2D"
                              Name 42  "usamp3D"
                              Name 49  "sampCube"
                              Name 56  "isamp1DA"
                              Name 63  "usamp2DA"
                              Name 70  "isampCubeA"
                              Name 77  "samp1Ds"
                              Name 84  "samp2Ds"
                              Name 91  "sampCubes"
                              Name 98  "samp1DAs"
                              Name 105  "samp2DAs"
                              Name 112  "sampCubeAs"
                              Name 116  "samp1D_iden"
                              Name 118  "isamp2D_iden"
                              Name 120  "usamp3D_iden"
                              Name 122  "sampCube_iden"
                              Name 124  "isamp1DA_iden"
                              Name 126  "usamp2DA_iden"
                              Name 128  "isampCubeA_iden"
                              Name 130  "samp1Ds_iden"
                              Name 132  "samp2Ds_iden"
                              Name 134  "sampCubes_iden"
                              Name 136  "samp1DAs_iden"
                              Name 138  "samp2DAs_iden"
                              Name 140  "sampCubeAs_iden"
                              Name 144  "i1D"
                              Name 147  "UBO"
                              MemberName 147(UBO) 0  "some_int"
                              MemberName 147(UBO) 1  "fixed_array"
                              MemberName 147(UBO) 2  "uboSampler"
                              MemberName 147(UBO) 3  "uboImage"
                              MemberName 147(UBO) 4  "textureHandle"
                              MemberName 147(UBO) 5  "textureHandleI"
                              Name 149  ""
                              Name 154  "i2D"
                              Name 160  "i3D"
                              Name 166  "iCube"
                              Name 172  "iCubeArray"
                              Name 178  "i1DArray"
                              Name 184  "i2DArray"
                              Name 190  "iBuffer"
                              Name 196  "i2DMS"
                              Name 202  "i2DMSArray"
                              Name 208  "ii1D"
                              Name 214  "ui2D"
                              Name 220  "ii2DMS"
                              Name 226  "ui2DMSArray"
                              Name 230  "i1D_iden"
                              Name 232  "i2D_iden"
                              Name 234  "i3D_iden"
                              Name 236  "iCube_iden"
                              Name 238  "iCubeArray_iden"
                              Name 240  "i1DArray_iden"
                              Name 242  "i2DArray_iden"
                              Name 244  "iBuffer_iden"
                              Name 246  "i2DMS_iden"
                              Name 248  "i2DMSArray_iden"
                              Name 250  "ii1D_iden"
                              Name 252  "ui2D_iden"
                              Name 254  "ii2DMS_iden"
                              Name 256  "ui2DMSArray_iden"
                              Name 259  "theLocalSampler"
                              Name 264  "theSampler"
                              Name 271  "localimg"
                              Name 276  "Light"
                              MemberName 276(Light) 0  "theStructSampler"
                              MemberName 276(Light) 1  "inImage"
                              Name 278  "test"
                              Name 287  "res"
                              Name 295  "res1"
                              Name 301  "res2"
                              Name 306  "utheSampler"
                              Name 310  "ulocalImage"
                              Name 315  "lu64utextureHandle"
                              Name 321  "lu64imgutextureHandle"
                              Name 330  "theImage1"
                              Name 332  "tCoord"
                              Decorate 18 ArrayStride 4
                              Decorate 21 ArrayStride 16
                              MemberDecorate 24(SSBO) 0 Offset 0
                              MemberDecorate 24(SSBO) 1 Offset 4
                              MemberDecorate 24(SSBO) 2 Offset 16
                              MemberDecorate 24(SSBO) 3 Offset 64
                              MemberDecorate 24(SSBO) 4 Offset 72
                              Decorate 24(SSBO) BufferBlock
                              Decorate 26 DescriptorSet 0
                              Decorate 26 Binding 0
                              Decorate 145 ArrayStride 16
                              Decorate 146 ArrayStride 16
                              MemberDecorate 147(UBO) 0 Offset 0
                              MemberDecorate 147(UBO) 1 Offset 16
                              MemberDecorate 147(UBO) 2 Offset 64
                              MemberDecorate 147(UBO) 3 Offset 208
                              MemberDecorate 147(UBO) 4 Offset 216
                              MemberDecorate 147(UBO) 5 Offset 224
                              Decorate 147(UBO) Block
                              Decorate 149 DescriptorSet 0
                              Decorate 149 Binding 2
                              Decorate 264(theSampler) Location 0
                              Decorate 264(theSampler) DescriptorSet 0
                              Decorate 264(theSampler) Binding 1
                              Decorate 264(theSampler) BindlessSamplerNV
                              Decorate 330(theImage1) Location 1
                              Decorate 330(theImage1) DescriptorSet 0
                              Decorate 330(theImage1) Binding 0
                              Decorate 330(theImage1) BindlessmageNV
                              Decorate 332(tCoord) Location 0
               2:             TypeVoid
               3:             TypeFunction 2
              10:             TypeFloat 32
              11:             TypeImage 10(float) 1D sampled format:Unknown
              12:             TypeSampledImage 11
              13:             TypePointer Function 12
              15:             TypeInt 32 1
              16:             TypeInt 32 0
              17:     16(int) Constant 3
              18:             TypeArray 10(float) 17
              19:             TypeImage 10(float) 2D sampled format:Unknown
              20:             TypeSampledImage 19
              21:             TypeArray 20 17
              22:             TypeImage 10(float) 2D nonsampled format:Rgba32f
              23:             TypeInt 64 0
        24(SSBO):             TypeStruct 15(int) 18 21 22 23(int64_t)
              25:             TypePointer Uniform 24(SSBO)
              26:     25(ptr) Variable Uniform
              27:     15(int) Constant 4
              28:             TypePointer Uniform 23(int64_t)
              32:             TypeImage 15(int) 2D sampled format:Unknown
              33:             TypeSampledImage 32
              34:             TypePointer Function 33
              39:             TypeImage 16(int) 3D sampled format:Unknown
              40:             TypeSampledImage 39
              41:             TypePointer Function 40
              46:             TypeImage 10(float) Cube sampled format:Unknown
              47:             TypeSampledImage 46
              48:             TypePointer Function 47
              53:             TypeImage 15(int) 1D array sampled format:Unknown
              54:             TypeSampledImage 53
              55:             TypePointer Function 54
              60:             TypeImage 16(int) 2D array sampled format:Unknown
              61:             TypeSampledImage 60
              62:             TypePointer Function 61
              67:             TypeImage 15(int) Cube array sampled format:Unknown
              68:             TypeSampledImage 67
              69:             TypePointer Function 68
              74:             TypeImage 10(float) 1D depth sampled format:Unknown
              75:             TypeSampledImage 74
              76:             TypePointer Function 75
              81:             TypeImage 10(float) 2D depth sampled format:Unknown
              82:             TypeSampledImage 81
              83:             TypePointer Function 82
              88:             TypeImage 10(float) Cube depth sampled format:Unknown
              89:             TypeSampledImage 88
              90:             TypePointer Function 89
              95:             TypeImage 10(float) 1D depth array sampled format:Unknown
              96:             TypeSampledImage 95
              97:             TypePointer Function 96
             102:             TypeImage 10(float) 2D depth array sampled format:Unknown
             103:             TypeSampledImage 102
             104:             TypePointer Function 103
             109:             TypeImage 10(float) Cube depth array sampled format:Unknown
             110:             TypeSampledImage 109
             111:             TypePointer Function 110
             142:             TypeImage 10(float) 1D nonsampled format:Rgba32f
             143:             TypePointer Function 142
             145:             TypeArray 10(float) 17
             146:             TypeArray 20 17
        147(UBO):             TypeStruct 15(int) 145 146 22 23(int64_t) 23(int64_t)
             148:             TypePointer Uniform 147(UBO)
             149:    148(ptr) Variable Uniform
             153:             TypePointer Function 22
             158:             TypeImage 10(float) 3D nonsampled format:Rgba32f
             159:             TypePointer Function 158
             164:             TypeImage 10(float) Cube nonsampled format:Rgba32f
             165:             TypePointer Function 164
             170:             TypeImage 10(float) Cube array nonsampled format:Rgba32f
             171:             TypePointer Function 170
             176:             TypeImage 10(float) 1D array nonsampled format:Rgba32f
             177:             TypePointer Function 176
             182:             TypeImage 10(float) 2D array nonsampled format:Rg16
             183:             TypePointer Function 182
             188:             TypeImage 10(float) Buffer nonsampled format:Rgba32f
             189:             TypePointer Function 188
             194:             TypeImage 10(float) 2D multi-sampled nonsampled format:Rgba32f
             195:             TypePointer Function 194
             200:             TypeImage 10(float) 2D array multi-sampled nonsampled format:Rgba32f
             201:             TypePointer Function 200
             206:             TypeImage 15(int) 1D nonsampled format:R32i
             207:             TypePointer Function 206
             212:             TypeImage 16(int) 2D nonsampled format:R32ui
             213:             TypePointer Function 212
             218:             TypeImage 15(int) 2D multi-sampled nonsampled format:R32i
             219:             TypePointer Function 218
             224:             TypeImage 16(int) 2D array multi-sampled nonsampled format:R32ui
             225:             TypePointer Function 224
             258:             TypePointer Function 20
             263:             TypePointer UniformConstant 20
 264(theSampler):    263(ptr) Variable UniformConstant
             272:     15(int) Constant 5
      276(Light):             TypeStruct 20 22
             277:             TypePointer Private 276(Light)
       278(test):    277(ptr) Variable Private
             279:     15(int) Constant 1
             283:             TypePointer Private 22
             285:             TypeVector 10(float) 4
             286:             TypePointer Function 285(fvec4)
             291:             TypeVector 10(float) 2
             292:   10(float) Constant 0
             293:  291(fvec2) ConstantComposite 292 292
             297:             TypeVector 15(int) 2
             298:     15(int) Constant 0
             299:  297(ivec2) ConstantComposite 298 298
             314:             TypePointer Function 23(int64_t)
             316:     15(int) Constant 2
             317:             TypePointer Uniform 20
             322:     15(int) Constant 3
             323:             TypePointer Uniform 22
             329:             TypePointer UniformConstant 22
  330(theImage1):    329(ptr) Variable UniformConstant
             331:             TypePointer Input 291(fvec2)
     332(tCoord):    331(ptr) Variable Input
         4(main):           2 Function None 3
               5:             Label
259(theLocalSampler):    258(ptr) Variable Function
   271(localimg):    153(ptr) Variable Function
        287(res):    286(ptr) Variable Function
       295(res1):    286(ptr) Variable Function
       301(res2):    286(ptr) Variable Function
306(utheSampler):    258(ptr) Variable Function
310(ulocalImage):    153(ptr) Variable Function
315(lu64utextureHandle):    314(ptr) Variable Function
321(lu64imgutextureHandle):    314(ptr) Variable Function
             260:     28(ptr) AccessChain 149 27
             261: 23(int64_t) Load 260
             262:          20 ConvertUToSampledImageNV 261
                              Store 259(theLocalSampler) 262
             265:          20 Load 264(theSampler)
             266:          19 Image 265
             267:     15(int) ImageQueryLevels 266
             268:          20 Load 259(theLocalSampler)
             269:          19 Image 268
             270:     15(int) ImageQueryLevels 269
             273:     28(ptr) AccessChain 149 272
             274: 23(int64_t) Load 273
             275:          22 ConvertUToImageNV 274
                              Store 271(localimg) 275
             280:     28(ptr) AccessChain 149 272
             281: 23(int64_t) Load 280
             282:          22 ConvertUToImageNV 281
             284:    283(ptr) AccessChain 278(test) 279
                              Store 284 282
             288:     28(ptr) AccessChain 149 27
             289: 23(int64_t) Load 288
             290:          20 ConvertUToSampledImageNV 289
             294:  285(fvec4) ImageSampleImplicitLod 290 293
                              Store 287(res) 294
             296:          22 Load 271(localimg)
             300:  285(fvec4) ImageRead 296 299
                              Store 295(res1) 300
             302:     28(ptr) AccessChain 149 272
             303: 23(int64_t) Load 302
             304:          22 ConvertUToImageNV 303
             305:  285(fvec4) ImageRead 304 299
                              Store 301(res2) 305
             307:     28(ptr) AccessChain 149 27
             308: 23(int64_t) Load 307
             309:          20 ConvertUToSampledImageNV 308
                              Store 306(utheSampler) 309
             311:     28(ptr) AccessChain 149 27
             312: 23(int64_t) Load 311
             313:          22 ConvertUToImageNV 312
                              Store 310(ulocalImage) 313
             318:    317(ptr) AccessChain 149 316 298
             319:          20 Load 318
             320: 23(int64_t) ConvertSampledImageToUNV 319
                              Store 315(lu64utextureHandle) 320
             324:    323(ptr) AccessChain 149 322
             325:          22 Load 324
             326: 23(int64_t) ConvertImageToUNV 325
                              Store 321(lu64imgutextureHandle) 326
             327:           2 FunctionCall 6(testSamplerConstructors()
             328:           2 FunctionCall 8(testImageConstructors()
                              Return
                              FunctionEnd
6(testSamplerConstructors():           2 Function None 3
               7:             Label
      14(samp1D):     13(ptr) Variable Function
     35(isamp2D):     34(ptr) Variable Function
     42(usamp3D):     41(ptr) Variable Function
    49(sampCube):     48(ptr) Variable Function
    56(isamp1DA):     55(ptr) Variable Function
    63(usamp2DA):     62(ptr) Variable Function
  70(isampCubeA):     69(ptr) Variable Function
     77(samp1Ds):     76(ptr) Variable Function
     84(samp2Ds):     83(ptr) Variable Function
   91(sampCubes):     90(ptr) Variable Function
    98(samp1DAs):     97(ptr) Variable Function
   105(samp2DAs):    104(ptr) Variable Function
 112(sampCubeAs):    111(ptr) Variable Function
116(samp1D_iden):     13(ptr) Variable Function
118(isamp2D_iden):     34(ptr) Variable Function
120(usamp3D_iden):     41(ptr) Variable Function
122(sampCube_iden):     48(ptr) Variable Function
124(isamp1DA_iden):     55(ptr) Variable Function
126(usamp2DA_iden):     62(ptr) Variable Function
128(isampCubeA_iden):     69(ptr) Variable Function
130(samp1Ds_iden):     76(ptr) Variable Function
132(samp2Ds_iden):     83(ptr) Variable Function
134(sampCubes_iden):     90(ptr) Variable Function
136(samp1DAs_iden):     97(ptr) Variable Function
138(samp2DAs_iden):    104(ptr) Variable Function
140(sampCubeAs_iden):    111(ptr) Variable Function
              29:     28(ptr) AccessChain 26 27
              30: 23(int64_t) Load 29
              31:          12 ConvertUToSampledImageNV 30
                              Store 14(samp1D) 31
              36:     28(ptr) AccessChain 26 27
              37: 23(int64_t) Load 36
              38:          33 ConvertUToSampledImageNV 37
                              Store 35(isamp2D) 38
              43:     28(ptr) AccessChain 26 27
              44: 23(int64_t) Load 43
              45:          40 ConvertUToSampledImageNV 44
                              Store 42(usamp3D) 45
              50:     28(ptr) AccessChain 26 27
              51: 23(int64_t) Load 50
              52:          47 ConvertUToSampledImageNV 51
                              Store 49(sampCube) 52
              57:     28(ptr) AccessChain 26 27
              58: 23(int64_t) Load 57
              59:          54 ConvertUToSampledImageNV 58
                              Store 56(isamp1DA) 59
              64:     28(ptr) AccessChain 26 27
              65: 23(int64_t) Load 64
              66:          61 ConvertUToSampledImageNV 65
                              Store 63(usamp2DA) 66
              71:     28(ptr) AccessChain 26 27
              72: 23(int64_t) Load 71
              73:          68 ConvertUToSampledImageNV 72
                              Store 70(isampCubeA) 73
              78:     28(ptr) AccessChain 26 27
              79: 23(int64_t) Load 78
              80:          75 ConvertUToSampledImageNV 79
                              Store 77(samp1Ds) 80
              85:     28(ptr) AccessChain 26 27
              86: 23(int64_t) Load 85
              87:          82 ConvertUToSampledImageNV 86
                              Store 84(samp2Ds) 87
              92:     28(ptr) AccessChain 26 27
              93: 23(int64_t) Load 92
              94:          89 ConvertUToSampledImageNV 93
                              Store 91(sampCubes) 94
              99:     28(ptr) AccessChain 26 27
             100: 23(int64_t) Load 99
             101:          96 ConvertUToSampledImageNV 100
                              Store 98(samp1DAs) 101
             106:     28(ptr) AccessChain 26 27
             107: 23(int64_t) Load 106
             108:         103 ConvertUToSampledImageNV 107
                              Store 105(samp2DAs) 108
             113:     28(ptr) AccessChain 26 27
             114: 23(int64_t) Load 113
             115:         110 ConvertUToSampledImageNV 114
                              Store 112(sampCubeAs) 115
             117:          12 Load 14(samp1D)
                              Store 116(samp1D_iden) 117
             119:          33 Load 35(isamp2D)
                              Store 118(isamp2D_iden) 119
             121:          40 Load 42(usamp3D)
                              Store 120(usamp3D_iden) 121
             123:          47 Load 49(sampCube)
                              Store 122(sampCube_iden) 123
             125:          54 Load 56(isamp1DA)
                              Store 124(isamp1DA_iden) 125
             127:          61 Load 63(usamp2DA)
                              Store 126(usamp2DA_iden) 127
             129:          68 Load 70(isampCubeA)
                              Store 128(isampCubeA_iden) 129
             131:          75 Load 77(samp1Ds)
                              Store 130(samp1Ds_iden) 131
             133:          82 Load 84(samp2Ds)
                              Store 132(samp2Ds_iden) 133
             135:          89 Load 91(sampCubes)
                              Store 134(sampCubes_iden) 135
             137:          96 Load 98(samp1DAs)
                              Store 136(samp1DAs_iden) 137
             139:         103 Load 105(samp2DAs)
                              Store 138(samp2DAs_iden) 139
             141:         110 Load 112(sampCubeAs)
                              Store 140(sampCubeAs_iden) 141
                              Return
                              FunctionEnd
8(testImageConstructors():           2 Function None 3
               9:             Label
        144(i1D):    143(ptr) Variable Function
        154(i2D):    153(ptr) Variable Function
        160(i3D):    159(ptr) Variable Function
      166(iCube):    165(ptr) Variable Function
 172(iCubeArray):    171(ptr) Variable Function
   178(i1DArray):    177(ptr) Variable Function
   184(i2DArray):    183(ptr) Variable Function
    190(iBuffer):    189(ptr) Variable Function
      196(i2DMS):    195(ptr) Variable Function
 202(i2DMSArray):    201(ptr) Variable Function
       208(ii1D):    207(ptr) Variable Function
       214(ui2D):    213(ptr) Variable Function
     220(ii2DMS):    219(ptr) Variable Function
226(ui2DMSArray):    225(ptr) Variable Function
   230(i1D_iden):    143(ptr) Variable Function
   232(i2D_iden):    153(ptr) Variable Function
   234(i3D_iden):    159(ptr) Variable Function
 236(iCube_iden):    165(ptr) Variable Function
238(iCubeArray_iden):    171(ptr) Variable Function
240(i1DArray_iden):    177(ptr) Variable Function
242(i2DArray_iden):    183(ptr) Variable Function
244(iBuffer_iden):    189(ptr) Variable Function
 246(i2DMS_iden):    195(ptr) Variable Function
248(i2DMSArray_iden):    201(ptr) Variable Function
  250(ii1D_iden):    207(ptr) Variable Function
  252(ui2D_iden):    213(ptr) Variable Function
254(ii2DMS_iden):    219(ptr) Variable Function
256(ui2DMSArray_iden):    225(ptr) Variable Function
             150:     28(ptr) AccessChain 149 27
             151: 23(int64_t) Load 150
             152:         142 ConvertUToImageNV 151
                              Store 144(i1D) 152
             155:     28(ptr) AccessChain 149 27
             156: 23(int64_t) Load 155
             157:          22 ConvertUToImageNV 156
                              Store 154(i2D) 157
             161:     28(ptr) AccessChain 149 27
             162: 23(int64_t) Load 161
             163:         158 ConvertUToImageNV 162
                              Store 160(i3D) 163
             167:     28(ptr) AccessChain 149 27
             168: 23(int64_t) Load 167
             169:         164 ConvertUToImageNV 168
                              Store 166(iCube) 169
             173:     28(ptr) AccessChain 26 27
             174: 23(int64_t) Load 173
             175:         170 ConvertUToImageNV 174
                              Store 172(iCubeArray) 175
             179:     28(ptr) AccessChain 26 27
             180: 23(int64_t) Load 179
             181:         176 ConvertUToImageNV 180
                              Store 178(i1DArray) 181
             185:     28(ptr) AccessChain 26 27
             186: 23(int64_t) Load 185
             187:         182 ConvertUToImageNV 186
                              Store 184(i2DArray) 187
             191:     28(ptr) AccessChain 149 27
             192: 23(int64_t) Load 191
             193:         188 ConvertUToImageNV 192
                              Store 190(iBuffer) 193
             197:     28(ptr) AccessChain 149 27
             198: 23(int64_t) Load 197
             199:         194 ConvertUToImageNV 198
                              Store 196(i2DMS) 199
             203:     28(ptr) AccessChain 26 27
             204: 23(int64_t) Load 203
             205:         200 ConvertUToImageNV 204
                              Store 202(i2DMSArray) 205
             209:     28(ptr) AccessChain 149 27
             210: 23(int64_t) Load 209
             211:         206 ConvertUToImageNV 210
                              Store 208(ii1D) 211
             215:     28(ptr) AccessChain 26 27
             216: 23(int64_t) Load 215
             217:         212 ConvertUToImageNV 216
                              Store 214(ui2D) 217
             221:     28(ptr) AccessChain 149 27
             222: 23(int64_t) Load 221
             223:         218 ConvertUToImageNV 222
                              Store 220(ii2DMS) 223
             227:     28(ptr) AccessChain 26 27
             228: 23(int64_t) Load 227
             229:         224 ConvertUToImageNV 228
                              Store 226(ui2DMSArray) 229
             231:         142 Load 144(i1D)
                              Store 230(i1D_iden) 231
             233:          22 Load 154(i2D)
                              Store 232(i2D_iden) 233
             235:         158 Load 160(i3D)
                              Store 234(i3D_iden) 235
             237:         164 Load 166(iCube)
                              Store 236(iCube_iden) 237
             239:         170 Load 172(iCubeArray)
                              Store 238(iCubeArray_iden) 239
             241:         176 Load 178(i1DArray)
                              Store 240(i1DArray_iden) 241
             243:         182 Load 184(i2DArray)
                              Store 242(i2DArray_iden) 243
             245:         188 Load 190(iBuffer)
                              Store 244(iBuffer_iden) 245
             247:         194 Load 196(i2DMS)
                              Store 246(i2DMS_iden) 247
             249:         200 Load 202(i2DMSArray)
                              Store 248(i2DMSArray_iden) 249
             251:         206 Load 208(ii1D)
                              Store 250(ii1D_iden) 251
             253:         212 Load 214(ui2D)
                              Store 252(ui2D_iden) 253
             255:         218 Load 220(ii2DMS)
                              Store 254(ii2DMS_iden) 255
             257:         224 Load 226(ui2DMSArray)
                              Store 256(ui2DMSArray_iden) 257
                              Return
                              FunctionEnd
